lib/java/src/org/apache/thrift/async/TAsyncMethodCall.java

```diff
@@ -225,8 +231,13 @@ private void cleanUpAndFireCallback(SelectionKey key) {
     key.interestOps(0);
     // this ensures that the TAsyncMethod instance doesn't hang around
     key.attach(null);
```

Comment:
It turned out that it's one key per connection.
So indeed canceling here does not work at all, it spoils the connection after the first call.
Instead, cancel() should be called when the connection is closed but is not unless aborted by exceptions, so key objects are certainly leaking.

As it now seems a completely separate problem so I'll stop pulling it here and focus on this patch for now.
Can you explain as to why you are introducing try catch anyway ?
It's the last bit that stops me from going forward.